<a href="https://colab.research.google.com/github/SaraBCoutinho/Desafio/blob/main/Valorian.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

📓 NOTEBOOK DO DESAFIO
---------------------------------------------------------



**Objeto**: Este notebook contém o desenvolvimento da solução para o Desafio proposto.



Para desenvolver a solução seguindo um pipeline, este notebook esta dividido nos seguintes tópicos: 
*  Importação das bibliotecas
*  Coleta de Dados 
*  Entendimento dos dados e Pré-processamento dos Dados 

## 👨‍💻 Importação de bibliotecas

Nesta etapa estao as bibliotecas que foram usadas no projeto. Ao lado de cada uma, está explicito sua usabilidade nele em forma de comentário. 

In [ ]:
 import pandas as pd #dataframe
 import math #expoente
 import regex as re #manipulação de strings
 import decimal # manipulaçao de valores de notaçao cientifica
 import scipy.stats as stats #teste de distribuiçao dos dados 
 from datetime import datetime #timestamp
 from sklearn.preprocessing import LabelEncoder

In [371]:
import matplotlib.pyplot as plt


## 🎲 Coleta de Dados

Nesta etapa o arquivo precisou ser dezipado e foi armazenado em formato de dataframe. 

In [ ]:
#Dezipando o arquivo  
!unzip /content/drive/MyDrive/Candidaturas/coleta.txt.zip

#Criando dataframe para armazenar os dados
Data=pd.read_csv('/content/coleta.txt',sep=" ", header=None) #separador reconhecido foi o espaço em branco

Archive:  /content/drive/MyDrive/Candidaturas/coleta.txt.zip
replace coleta.txt? [y]es, [n]o, [A]ll, [N]one, [r]ename: n
replace __MACOSX/._coleta.txt? [y]es, [n]o, [A]ll, [N]one, [r]ename: n


## 🎲 Entendimento dos Dados e Pré-Processamento dos Dados

Esta etapa contou com aplicaçao dos conceitos Data Lateracy buscando o entendimento dos dados e analise sobre eles para poder tratá-los corretamente e aplicar corretamente os algoritmos de aprendizagem. Alguns pontos foram seguidos: 


1.   Perguntas sobre os dados
2.   Limpeza dos dados e eliminação de ruídos
3.   Compreender dados uteis
4.   Construir análises estatísticas

As perguntas que guiaram a resoluçao foram: 

*   Quantos dados existem? 

*   As unidades dos dados esta de acordo com as variavéis propostas?

*   Quais os ruidos existentes nos dados?

*   Cada coluna do dataframe corresponde a qual variavel?

Com base nisso foram cumpridos esses pontos. 


### Perguntas sobre os dados 

In [231]:
#Visualizando 5 primeiras linhas do dataframe
pd.DataFrame(Data.head(5))

0
0  33,Jogging,49105962326000,-0.6946377,12.680544...
1  33,Jogging,49106062271000,5.012288,11.264028,0...
2  33,Jogging,49106112167000,4.903325,10.882658,-...
3  33,Jogging,49106222305000,-0.61291564,18.49643...
4  33,Jogging,49106332290000,-1.1849703,12.108489...

In [232]:
#Shape do dataframe: (quantidade de instancias, quantidade de colunas)
print((Data.shape))

(1098204, 1)


Por enquanto existem 1098204 instâncias e apenas 1 coluna.  Verifica-se que os dados estão num formato que precisa de ajuste.  Desta forma serão separados e cada elemento entre as virgulas será campo de nova coluna. De acordo com os campos, as colunas constatadas #df=df.rename(columns={0:'Individuo',1:'Atividade',2:'Timestamp',3:'X',4:'Y',5:'Z',7:'Atividade',8:'Timestamp',9:'X',10:'Y',11:'Z'})

In [279]:
#Separando os campos por virgula da coluna existente cujo nome é 0 
Data[0] = Data[0].str.split(',')

Cada elemento está em uma lista. Serão separados em diferentes colunas. Serão verificadas quantas colunas existem, pois podem ser mais do que a quantidade de variáveis propostas ou menos. 

In [234]:
maximo = 0
for i in range(len(Data[0])):
  tamanho=len(Data[0].iloc[i])
  if tamanho > maximo:
    maximo = tamanho
print(maximo)

11


O máximo de colunas é 11. Elas ja serao divididas nesse topico para que o conjunto de dados possa ser compreendido melhor. Será verificado o que cada coluna significa, os valores contidos nelas e o tipo de dado de cada uma. 

In [280]:
Data = pd.DataFrame(Data[0].to_list(), columns=[0,1,2,3,4,5,6,7,8,9,10])

In [238]:
#Visualizando a descrição dos dados
pd.DataFrame(Data.describe())

0        1        2        3        4        5      6  7  8  9   \
count   1098204  1098204  1098204  1098204  1098204  1098204  11738  5  5  5   
unique       36        6  1006170     1856     1812     2681      3  4  4  4   
top          20  Walking        0        0    19.57     0.0;      ;  0  0  0   
freq      56632   424398    12843    16444    16345    12822  11733  2  2  2   

          10  
count      5  
unique     4  
top     0.0;  
freq       2

In [239]:
#Visualizando 1 instância dos dados
pd.DataFrame(Data.head(1))

0        1               2           3          4            5     6   \
0  33  Jogging  49105962326000  -0.6946377  12.680544  0.50395286;  None   

     7     8     9     10  
0  None  None  None  None

In [243]:
#Visualizando valores unicos das colunas que tem menos valores nulos pela descriçao
print('Valores únicos coluna 0:', Data[0].unique())
print('Valores únicos coluna 1:', Data[1].unique())
print('Valores únicos coluna 3:', Data[3].unique())
print('Valores únicos coluna 4:', Data[4].unique())
print('Valores únicos coluna 5:', Data[5].unique())

Valores únicos coluna 0: ['33' '17' '20' '29' '13' '15' '6' '27' '36' '18' '32' '35' '11' '16' '5'
 '10' '28' '26' '14' '24' '12' '23' '4' '30' '34' '8' '31' '21' '3' '22'
 '1' '9' '25' '2' '7' '19']
Valores únicos coluna 1: ['Jogging' 'Walking' 'Upstairs' 'Downstairs' 'Sitting' 'Standing']
Valores únicos coluna 3: ['-0.6946377' '5.012288' '4.903325' ... '19.65' '19.91' '19.95']
Valores únicos coluna 4: ['12.680544' '11.264028' '10.882658' ... '20.04' '19.8' '19.89']
Valores únicos coluna 5: ['0.50395286;' '0.95342433;' '-0.08172209;' ... '-19.76;' '-19.8;'
 '-19.7;']


In [240]:
#Visualizando valores unicos das colunas que tem mais valores nulos pela descriçao
print('Valores únicos coluna 6:', Data[6].unique())
print('Valores únicos coluna 7:', Data[7].unique())
print('Valores únicos coluna 8:', Data[8].unique())
print('Valores únicos coluna 9:', Data[9].unique())
print('Valores únicos coluna 10:', Data[10].unique())

Valores únicos coluna 6: [None 'Jogging' 'Walking' ';']
Valores únicos coluna 7: [None '328982249000' '0' '1202262279000' '295812327000']
Valores únicos coluna 8: [None '5.37' '0' '-1.23' '0.8']
Valores únicos coluna 9: [None '-3.36' '0' '3.45' '18.96']
Valores únicos coluna 10: [None '7.3141265;' '0.0;' '-0.38136974;' '13.57;']


In [246]:
#Visualizando a quantidade das açoes da coluna 6 que parecem ser consecutivas a primeira açao 
print(Data[Data[6] == ';'].shape)
print(Data[Data[6] == 'Jogging'].shape)
print(Data[Data[6] == 'Walking'].shape)

(11733, 11)
(3, 11)
(2, 11)


In [258]:
#Visualizando açoes com ';'
Data[Data[6] == ';'][[7,8,9,10]].describe()

7    8    9    10
count     0    0    0    0
unique    0    0    0    0
top     NaN  NaN  NaN  NaN
freq    NaN  NaN  NaN  NaN

In [259]:
#Verificando o tipo de dado de cada coluna
Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1098204 entries, 0 to 1098203
Data columns (total 11 columns):
 #   Column  Non-Null Count    Dtype 
---  ------  --------------    ----- 
 0   0       1098204 non-null  object
 1   1       1098204 non-null  object
 2   2       1098204 non-null  object
 3   3       1098204 non-null  object
 4   4       1098204 non-null  object
 5   5       1098204 non-null  object
 6   6       11738 non-null    object
 7   7       5 non-null        object
 8   8       5 non-null        object
 9   9       5 non-null        object
 10  10      5 non-null        object
dtypes: object(11)
memory usage: 92.2+ MB


Observaçoes obtidas sobre os campos e o entendimento do que representa provavelmente cada um, de acordo com a descriçao do desafio:

*   0 é o individuo - Existem 36 indivíduos
*   1 e 6 a atividade - Existem 6 ações, na coluna 6 precisa ser retirado o ';' que foi uma falha no registro apenas e nao significa uma outra açao pois as colunas 7,8,9 e 10 nao trazem nada. Além disso, sera considerada também a açao de nao faze nada, que corresponde ao campo nulo da coluna 6. 
*   2 e o 7 o timestamp - Esta em Nanossegundo e precisa vir pra segundo, na coluna 7 o campo 0 precisa ser transformado de string pra numero
*   trio 3, 4 e 5  e o 8, 9 e 10 as acelerações de X, Y e Z cada um respectivamente - Verificar a quantidade de casas decimais. Para o 8, 9 e 10 provavelmente o indivíduo realizou ações consecutivas e essas colunas não devem ser apagadas mas representam uma informação importante.

A seguir algumas observaçoes que serao necessarias para cada etapa: 

*   Limpeza: Retirar duplicatas, preencher valores ausentes pode ser 0 ja que nenhuma açao gera nenhuma aceleraçao nos campos dos registros dos sensores
*   Pré-processamento: Ajeitar valores das colunas 5,6 e 10 que alguns vem com o caracter ';' , categorizar as colunas 1 e 6, transformar coluna 7 o valor 0 de string pra numero e transformar de nanossegundo para segundo nas colunas 7 e 2, transformar os dados pra numéricos
*   Os algoritmos demandados requerem que seja feita normalizaçao, sobretudo uma vez que o XGBoost sera aplicado e tem por base o conceito de gradient descendent. Contudo, essa sera aplicada junto ao classificador a partir da biblioteca pipeline. 



#### Limpeza dos Dados 

In [281]:
# Retirada de duplicatas caso existam
print(Data.shape) #antigo tamanho do dataframe
print(Data.duplicated().sum())
Data = Data.drop_duplicates()
print(Data.shape) #novo tamanho do dataframe

(1098204, 11)
66817
(1031387, 11)


Existem agora 1031387 instâncias. 

In [262]:
#Preenchimento de valores ausentes colunas 0 a 5
print(Data[[0,1,2,3,4,5]].isnull().sum())

0    0
1    0
2    0
3    0
4    0
5    0
dtype: int64


In [282]:
#Preenchimento de valores ausentes colunas 6 a 10
Data[6]=Data[6].fillna(0)
Data[7]=Data[7].fillna(0) 
Data[8]=Data[8].fillna(0)
Data[9]=Data[9].fillna(0)
Data[10]=Data[10].fillna(0)

In [263]:
#Verificaçao do Preenchimento de valores ausentes 
print(Data[[6,7,8,9,10]].isnull().sum())

6     0
7     0
8     0
9     0
10    0
dtype: int64


#### Pré-processamento 

In [272]:
#Para a atividade de Substituição das variaveis categoricas
le = LabelEncoder()

In [283]:
#Corrigindo valores colunas 5,6 e 10
Data[6] = Data[6].replace(';','0')
Data[6] = Data[6].replace(0,'0')

In [284]:
print(Data[5].unique())
print(Data[10].unique())

['0.50395286;' '0.95342433;' '-0.08172209;' ... '-19.76;' '-19.8;'
 '-19.7;']
[0 '7.3141265;' '0.0;' '-0.38136974;' '13.57;']


In [290]:
#Substituindo valores da Coluna 5
print(Data[Data[5].str.contains(";")].shape) #verificando quantos valores tem ';' nessa coluna

def Remove_string(string): #funçao para apagar ';' da string
  return  re.sub(';','',string)

Data[5] = Data[5].map(Remove_string)

print(Data[Data[5].str.contains(";")].shape) #verificando se ';' foi retirado de todas

(1019883, 11)
(0, 11)


In [292]:
#Substituindo valores da Coluna 10
Substitute_string={0:'0.0','7.3141265;':'7.3141265','0.0;':'0.0','-0.38136974;':'-0.38136974','13.57;':'13.57'} #função para substituir os valores
Data[10]=Data[10].map(Substitute_string) #aplicação da função
Data[Data[10].str.contains(";")].shape #verificação se os valores foram substituidos

(0, 11)

In [293]:
#Categorizando colunas 1 e 6 
Data[1] = le.fit_transform(Data[1])
Data[6] = le.fit_transform(Data[6])

Para a transformaçao de Nanossegundo pra segundo, a variaçao sao de 1e-09 (10 elevado a 9) casas. Adotou-se por abordagem a retirada dos 9 ultimos elementos e mantida a parte inteira desse dado, o que corresponde aos segundos. 

In [207]:
#Corrigindo coluna 7 
[7]=Data[7].fillna(0) 

In [298]:
Data[7].unique()

array([0, '328982249000', '0', '1202262279000', '295812327000'],
      dtype=object)

In [299]:
Change_time_7={0:'0','0':'0','328982249000':'328','1202262279000':'12022','295812327000':'2958'}
Data[7]=Data[7].map(Change_time_7)

#Corrigindo a coluna 2
def Change_time_2(item): #funçao para correçao da coluna 2
  size = len(item)
  return item[:size - 9]
Data[2] = Data[2].map(Change_time_2)

In [304]:
#Transformaçao dos dados para numerico

#Novamente verificando o tipo de dado
Data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1031387 entries, 0 to 1098203
Data columns (total 11 columns):
 #   Column  Non-Null Count    Dtype 
---  ------  --------------    ----- 
 0   0       1031387 non-null  object
 1   1       1031387 non-null  int64 
 2   2       1031387 non-null  object
 3   3       1031387 non-null  object
 4   4       1031387 non-null  object
 5   5       1031387 non-null  object
 6   6       1031387 non-null  int64 
 7   7       1031387 non-null  object
 8   8       1031387 non-null  object
 9   9       1031387 non-null  object
 10  10      1031387 non-null  object
dtypes: int64(2), object(9)
memory usage: 94.4+ MB


As correçoes serao feitas para os que ainda nao sao numéricos.

In [305]:
#convertendo os tipos de variaveis
## individuo
Data[0] = Data[0].apply(int)

## timestamp
Data[2] = pd.to_numeric(Data[2])
Data[7] = pd.to_numeric(Data[7])

## acelerações X, Y e Z
Data[3] = Data[3].apply(float)
Data[4] = Data[4].apply(float)
Data[5] = pd.to_numeric(Data[5])

Data[8] = Data[8].apply(float)
Data[9] = Data[9].apply(float)
Data[10] = Data[10].apply(float)

In [308]:
#Verificando mais uma vez os dados estao numericos
Data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1031387 entries, 0 to 1098203
Data columns (total 11 columns):
 #   Column  Non-Null Count    Dtype  
---  ------  --------------    -----  
 0   0       1031387 non-null  int64  
 1   1       1031387 non-null  int64  
 2   2       1031343 non-null  float64
 3   3       1031387 non-null  float64
 4   4       1031387 non-null  float64
 5   5       1031386 non-null  float64
 6   6       1031387 non-null  int64  
 7   7       1031387 non-null  int64  
 8   8       1031387 non-null  float64
 9   9       1031387 non-null  float64
 10  10      1031387 non-null  float64
dtypes: float64(7), int64(4)
memory usage: 94.4 MB


In [309]:
#Retirada de valores nulos pela tabela
Data.dropna(inplace=True)

In [310]:
#Verificando mais uma vez os dados estao nao nulos
Data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1031342 entries, 0 to 1098203
Data columns (total 11 columns):
 #   Column  Non-Null Count    Dtype  
---  ------  --------------    -----  
 0   0       1031342 non-null  int64  
 1   1       1031342 non-null  int64  
 2   2       1031342 non-null  float64
 3   3       1031342 non-null  float64
 4   4       1031342 non-null  float64
 5   5       1031342 non-null  float64
 6   6       1031342 non-null  int64  
 7   7       1031342 non-null  int64  
 8   8       1031342 non-null  float64
 9   9       1031342 non-null  float64
 10  10      1031342 non-null  float64
dtypes: float64(7), int64(4)
memory usage: 94.4 MB


Os dados sao salvos no arquivo abaixo para a parte da detecçao das atividades. 

In [349]:
Data.to_csv('Dados_pre_processados.csv')